// instagrapi/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
var Utils = (function () {
    function Utils() {
    }
    Utils.msToDate = function (ms) {
        var _ms = ms / 1e10 < 1 ? ms * 1000 : ms;
        var date = new Date(_ms).toISOString();
        return date;
    };
    Utils.getPostUrl = function (code) {
        return global.config.urlBase + "/p/" + code;
    };
    Utils.getCaption = function (media) {
        var caption = media.edge_media_to_caption.edges.length
            ? media.edge_media_to_caption.edges[0].node.text
            : media.accessibility_caption || '';
        return caption ? caption : media.is_video ? media.title : null;
    };
    Utils.getComments = function (commentList) {
        if (commentList === null || commentList === void 0 ? void 0 : commentList.length) {
            return commentList.map(function (_a) {
                var _b;
                var comment = _a.node;
                var user = comment.owner;
                var commentList = (_b = comment.edge_threaded_comments) === null || _b === void 0 ? void 0 : _b.edges;
                return {
                    content: comment.text,
                    author: {
                        username: user.username,
                        image: user.profile_pic_url,
                        isVerified: user.is_verified
                    },
                    likes: comment.edge_liked_by.count,
                    responses: Utils.getComments(commentList),
                    isSpam: comment.did_report_as_spam,
                    date: Utils.msToDate(comment.created_at)
                };
            });
        }
        else
            return [];
    };
    Utils.getLocation = function (addressJson) {
        var address = JSON.parse(addressJson);
        var location = {
            country: address.country_code || null,
            region: address.region_name || null,
            city: address.city_name || null,
            street: address.street_address || null,
            zipCode: address.zip_code || null
        };
        return location;
    };
    return Utils;
}());
exports.Utils = Utils;
